op-chain-ops/cmd/celo-migrate/state.go

```diff
@@ -242,48 +246,72 @@ func applyStateMigrationChanges(config *genesis.DeployConfig, genesis *core.Gene
 // If an account already exists, it adds the balance of the new account to the existing balance.
 // If the code of an existing account is different from the code in the genesis block, it logs a warning.
 // This changes the state root, so `Commit` needs to be called after this function.
-func applyAllocsToState(db *state.StateDB, genesis *core.Genesis, config *params.ChainConfig) {
+func applyAllocsToState(db vm.StateDB, genesis *core.Genesis, config *params.ChainConfig) error {
 	log.Info("Starting to migrate OP contracts into state DB")
 
-	accountCounter := 0
+	copyCounter := 0
+	skipCounter := 0
 	overwriteCounter := 0
-	for k, v := range genesis.Alloc {
-		accountCounter++
+	whitelist := accountOverwriteWhitelist[config.ChainID.Uint64()]
 
-		balance := uint256.MustFromBig(v.Balance)
+	for k, v := range genesis.Alloc {
+		// Check that the balance of the account to written is zero,
+		// as we must not create new CELo tokens
+		if v.Balance.Cmp(big.NewInt(0)) != 0 {
+			log.Error("Account balance is not zero, would change celo supply", "address", k.Hex())
+			return fmt.Errorf("account balance is not zero, would change celo supply: %s", k.Hex())
+		}
 
 		if db.Exist(k) {
-			// If the account already has balance, add it to the balance of the new account
-			balance = balance.Add(balance, db.GetBalance(k))
-
-			currentCode := db.GetCode(k)
-			equalCode := bytes.Equal(currentCode, v.Code)
-			if currentCode != nil && !equalCode {
-				if whitelist, exists := accountOverwriteWhitelist[config.ChainID.Uint64()]; exists {
+			if db.GetCodeSize(k) == 0 {
+				// Check if it's an EOA, if so bail out
+				if db.GetNonce(k) > 0 {
+					log.Error("Account already exists with nonce > 0", "address", k.Hex())
+					return fmt.Errorf("account already exists with nonce > 0: %s", k.Hex())
+				} else {
+					copyAlloc(db, k, v, v.Nonce)
+					copyCounter++
+				}
+			} else { // account has code
+				equalCode := bytes.Equal(db.GetCode(k), v.Code)
+				if equalCode {
+					log.Info("Account already exists with same code, skipping...", "address", k.Hex())
+					// No need to overwrite the account, keep existing state and nonce
```

Comment:
Looking into the allocs I can see that we have accounts with nonces of 0 or 1 and also some have state.

So I don't see why it would be safe to skip updating the nonce and state in this situation. If some other part of the system is relying on the nonce and state being set to certain values then skipping here is going to break something.

Reply:
You can also argue the other direction.

In this case there's the exact contract already deployed and we don't want to destroy any other user relying on the existing state. So we shouldn't overwrite state or nonce.

Reply:
To conclude, I think this case is behaviour where we need to look at the particular case.

Reply:
I think in this case we need to fail and then manually asses the exact situation. I don't think we can safely just choose one or other approaches.

The nonce for a smart contract address is used to derive the addresses of contracts deployed by the smart contract, so having a different nonce may cause other contracts to not be able to find the contract they are looking for, or to mistakenly use a different deployment than the one they wanted or for the deployer to end up stuck unable to deploy any contracts.

Likewise I assume the state in the allocs is important, we can't just ignore it.

Reply:
Ok, I made it fail in that case.